### PR TITLE
[java] put function meta in task spec and load functions with function meta

### PR DESCRIPTION
**Pull request by raulchen**
This PR adds a `function_desc` field into task spec. a function descriptor is a list of strings that can uniquely describe a function.
- For a Python function, it should be: [module_name, class_name, function_name]
- For a Java function, it should be: [class_name, method_name, type_descriptor]

There're a couple of purposes to add this field:

In this PR:
- Java worker needs to know function's class name to load it. Previously, since task spec didn't have such a field to hold this info, we did a hack by appending the class name to the argument list. With this change, we fixed that hack and significantly simplified function management in Java.

Will be done in subsequent PRs:
- Support cross-language invocation (#2576): currently Python worker manages functions by saving them in GCS and pass function id in task spec. However, if we want to call a Python function from Java, we cannot save it in GCS and get the function id. But instead, we can pass the function descriptor (module name, class name, function name) in task spec and use it to load the function.
- Support deployment: one major problem of Python worker's current function management mechanism is #2327. In prod env, we should have a mechanism to deploy code and dependencies to the cluster. And when code is already deployed, we don't need to save functions to GCS any more and can use `function_desc` to manage functions.

### CODE REVIEW

**Comment by robertnishihara** [comment]
A couple questions,

1. How would you handle the situation
    ```python
    @ray.remote
    def f():
        return 1

    @ray.remote
    def f():
        return 2
    ```
    In this case the function name + module does not uniquely identify the function.
2. The idea that the "function description" uniquely identifies the function feels redundant with the "function ID". Do we really need both? Should we get rid of one? Maybe we don't actually need function IDs.
3. For cross-language deployment, we may need to store some sort mapping from "function module + name" to "function ID", but this could be done in a GCS table and doesn't need to be in every task specification, right?

**Comment by raulchen** [comment]
1) In Python, if you define 2 functions with the same name in one module, the second one will overwrite the first. That means, `ray.get(f.remote())` will always return 2, regardless of this change. 
2) yeah, we should also change python worker to use function_desc for function management, and then we can remove function_id. Note that we just use function_desc as the key to look up functions, we can still store pickled functions in GCS if we don't deploy code. 
3) I don't understand why we need to store that mapping. Could you explain why? I think, with function_desc, we should have enough info to load a function.

**Comment by robertnishihara** [comment]
I'm a bit skeptical about your answer to 1. Someone may still have a reference to the original function, or tasks may be submitted between the two definitions.

**Comment by raulchen** [comment]
oh, I just found that currently the function_id is just the hash of module name and function name, in non-script case. So I think the behavior won't change. see https://github.com/ray-project/ray/blob/b7722897b4c341a28002a917ef1d136458f4beec/python/ray/remote_function.py#L18

**Comment by raulchen** [comment]
@robertnishihara, did my comments answer your questions?

**Comment by robertnishihara** [comment]
@raulchen we try to hash the source code into the function ID as well, though it is imperfect. This has caused issues like https://github.com/ray-project/ray/issues/2732. One solution may be to make the function ID include a hash of the pickled function itself.

**Comment by robertnishihara** [comment]
But yes, I do think this scenario matters, especially because it happens all the time in the shell, and also people may define a function inside of a loop.

**Comment by robertnishihara** [comment]
Can you say a bit about how the worker would use this information?

One option is for the Java worker that creates the task to compute (or look up) the correct function ID from the function name and module name and then simply put the function ID in the task spec.

**Comment by robertnishihara** [comment]
@pcmoritz can you take a look at this?

File: src/common/format/common.fbs [hunk 1]
```diff
@@ -62,6 +62,11 @@ table TaskInfo {
   required_resources: [ResourcePair];
   // The language that this task belongs to
   language: TaskLanguage;
+  // Function descriptor, which is a list of strings that can
+  // uniquely describe a function.
+  // For a Python function, it should be: [module_name, class_name, function_name]
+  // For a Java function, it should be: [class_name, method_name, type_descriptor]
+  function_desc: [string];
```
Comment (robertnishihara) [inline src/common/format/common.fbs]: `function_description` (please avoid abbreviations)

**Comment by raulchen** [comment]
@robertnishihara I've renamed `function_desc` to `function_descriptor`

**Review by robertnishihara — APPROVED** [review]

**Pull request merged**